### DISCUSSION TITLE
Saving exif metadata

### DISCUSSION THREAD

**Discussion by alandefreitas** [bigcat88/pillow_heif#25] [Q&A] [answered]
Hi!

Thanks for this library. It's the only way I'm being able to manipulate my images from iOS. :)

So... it seems like Digikam cannot update the dates in the exif metadata of these files, and this information only goes to the sidecar .xmp files. Then I've been trying to sync these metadata with something like:

```python
                image = PIL.Image.open(file)
                image.load()
                exifdata = image.getexif()
                datetime_tags = [
                    0x0132,  # "DateTime"
                    0x9003,  # "DateTimeOriginal"
                    0x9004,  # "DateTimeDigitized"
                    0xC71B,  # "PreviewDateTime"
                ]
                for tag_id in datetime_tags:
                    # new_exif_str is a date string in the '%Y:%m:%d %H:%M:%S' format
                    exifdata[tag_id] = new_exif_str
                image.save(file, "heif", exif=exifdata.tobytes())
```

When I open the file, the image has changed (it has been compressed, etc), but the metadata is exactly the same.

What am I doing wrong?

**Comment by bigcat88** [comment, marked as answer]
I dint write support for this:
`image.save(file, "heif", exif=exifdata.tobytes())`
Maybe I should do that....

`image.info["exif"] = exifdata.tobytes()`
`image.save(file)`
 should work, i will look at it...

**Comment by bigcat88** [comment]
I will add ability to specify "exif" and "xmp" as a keys for encoder to make your version of code work.

As for current code, version what I proposed works, but there is some problem for overriding all values for your tags in your code.
You can check this using this code:
```python3 
    image.load()
    exifdata = image.getexif()
    datetime_tags = [
        0x0132,  # "DateTime"
        0x9003,  # "DateTimeOriginal"
        0x9004,  # "DateTimeDigitized"
        0xC71B,  # "PreviewDateTime"
    ]
    for tag_id in datetime_tags:
        # new_exif_str is a date string in the '%Y:%m:%d %H:%M:%S' format
        exifdata[tag_id] = "1988:02:02 11:11:11"
    import piexif
    print("\tExif:")
    exif_dict = piexif.load(exifdata.tobytes(), key_is_name=True)
    for key, value in exif_dict.items():
        print(f"\t\t{key}:")
        if value is not None:
            for sub_key, sub_value in value.items():
                if isinstance(sub_value, bytes) and len(sub_value) > 20:
                    print(f"\t\t\t{sub_key}: {len(sub_value)} bytes.")
                else:
                    print(f"\t\t\t{sub_key}: {sub_value}")```

DateTimeOriginal doesn't not get overridden.

**Reply by alandefreitas** [reply]
Oh... I assumed "exif" metadata should work somehow because everything else works so perfectly. Good to know I was right.

I would never guess

```python
image.save(file, "heif", exif=exifdata.tobytes())
```

wouldn't work but 

```python
image.info["exif"] = exifdata.tobytes()
image.save(file)
```

That's great. It did work locally. 🥳

I'm not sure I understand the reason why `DateTimeOriginal` is not overridden in your second example. I've tested the image with [some other application](https://products.groupdocs.app/metadata/heic) and the attributes were fine:

![image](https://user-images.githubusercontent.com/5369819/171279417-def6e66d-bd48-4ced-b70b-6483dd4e9eaf.png)

They were some wrong dates (2021/something) before so it seems to be working. Is there a problem with the `exifdata[tag_id] = ...` assignment pattern above ? Is there a better pattern?

**Reply by bigcat88** [reply]
That's pattern looks fine, i am not so familiar with exif tags to advice something better.
Glad that you get worked what you need :)

**Reply by alandefreitas** [reply]
Thanks a lot!

Since I'm asking anyway (🙂), what about the Pillow `quality` argument in `Image::save`?

Because of compression, since I'll probably lose some image quality every time I save the file, is it possible to just update the metadata without touching with the image? 

If not, is there a pattern I can follow to maintain max image quality without writing a huge file? I tried `quality=100` but some files were about 4 times larger. I'm not comfortable converting the whole library at a lower quality either. I even considered a loop that would go saving new images until I find the smallest image not smaller than the original. 😆

**Comment by bigcat88** [comment]
> is it possible to just update the metadata without touching with the image?

Libheif itself doesn't support saving files in place :( 
So this is not possible, for now at least.

> If not, is there a pattern I can follow to maintain max image quality without writing a huge file? I tried quality=100 but some files were about 4 times larger. I'm not comfortable converting the whole library at a lower quality either. 

Next project I consider to develop, looks the same. 
Writing a `.heic` converter for Nextcloud to maintain photo library in HEIF format. :)

It is a big the problem, HEIF itself does not provide any information for encoding image in the same quality.
[libheif lossless encoding discussion](https://github.com/strukturag/libheif/issues/62#issuecomment-691129594)

> I even considered a loop that would go saving new images until I find the smallest image not smaller than the original.

You can laugh, but it seems the only available solution for now.
If you do that you can make a PR for such function for this project, it will be great. =)

**Reply by alandefreitas** [reply]
> Next project I consider to develop, looks the same.
> Writing a .heic converter for Nextcloud to maintain photo library in HEIF format. :)

So maybe that loop is not such a bad idea after all :)

The compression rates are pretty good. I would almost consider converting all the JPGs in my library to HEIC if the software I use to manage the photos weren't so slow with HEIC.

> It is a big the problem, HEIF itself does not provide any information for encoding image in the same quality.
> https://github.com/strukturag/libheif/issues/62#issuecomment-691129594

Isn't that problem different in that he wants to convert the image from PNG to HEIC, and back? Losing some quality is understandable since there are conversions involved. I just assumed updating the metadata and saving in the same format wouldn't be as difficult.

> You can laugh, but it seems the only available solution for now.

Oh... That's sad.

> If you do that you can make a PR for such function for this project, it will be great. =)

Sure. It's probably just a few lines of code. Code quality might not be great though. :)
